File: chapter-02/exercise-03/pointer_address_trace.cpp
#include <cassert>
#include <iostream>

int main() {
    int years = 0;
    int m = 0;
    double amt = 0.0;
    int firstnum = 154;
    int slope = 0;
    int k = 0;

    int* ptNum = nullptr;
    double* amtAddr = nullptr;
    int* zAddr = &slope;
    int* numAddr = &firstnum;
    int* ptDay = nullptr;
    int* ptYr = &years;

    ptNum = &m;
    amtAddr = &amt;
    *zAddr = 25;
    k = *numAddr;
    ptDay = zAddr;
    *ptYr = 1987;
    *amtAddr = *numAddr;

    assert(ptNum == &m);
    assert(amtAddr == &amt);
    assert(slope == 25);
    assert(k == 154);
    assert(ptDay == &slope);
    assert(years == 1987);
    assert(amt == 154.0);

    std::cout << "Illustrated final data from the presentation:\n";
    std::cout << "ptNum = 8096\n";
    std::cout << "amtAddr = 16256\n";
    std::cout << "zAddr = 20492\n";
    std::cout << "numAddr = 18938\n";
    std::cout << "ptDay = 20492\n";
    std::cout << "ptYr = 694\n";
    std::cout << "years = " << years << '\n';
    std::cout << "amt = " << amt << '\n';
    std::cout << "firstnum = " << firstnum << '\n';
    std::cout << "slope = " << slope << '\n';
    std::cout << "k = " << k << '\n';
}

File: chapter-02/exercise-03/pointer_address_trace.py
"""Simulate the presentation's pointer-and-address memory diagram in Python."""

ADDRESSES = {
    "ptNum": 500,
    "amtAddr": 564,
    "zAddr": 8024,
    "numAddr": 10132,
    "ptDay": 14862,
    "ptYr": 15010,
    "years": 694,
    "m": 8096,
    "amt": 16256,
    "firstnum": 18938,
    "slope": 20492,
    "k": 24608,
}


def run_trace() -> tuple[dict[int, int | None], dict[str, int]]:
    """Apply the seven statements using dictionaries as simulated memory."""
    memory: dict[int, int | None] = {
        694: None,
        8096: None,
        16256: None,
        18938: 154,
        20492: None,
        24608: None,
    }
    pointers = {
        "ptNum": 0,
        "amtAddr": 0,
        "zAddr": 20492,
        "numAddr": 18938,
        "ptDay": 0,
        "ptYr": 694,
    }

    pointers["ptNum"] = ADDRESSES["m"]
    pointers["amtAddr"] = ADDRESSES["amt"]
    memory[pointers["zAddr"]] = 25
    memory[ADDRESSES["k"]] = memory[pointers["numAddr"]]
    pointers["ptDay"] = pointers["zAddr"]
    memory[pointers["ptYr"]] = 1987
    memory[pointers["amtAddr"]] = memory[pointers["numAddr"]]

    return memory, pointers


def main() -> None:
    memory, pointers = run_trace()

    assert pointers["ptNum"] == 8096
    assert pointers["amtAddr"] == 16256
    assert pointers["ptDay"] == 20492
    assert memory[694] == 1987
    assert memory[16256] == 154
    assert memory[20492] == 25
    assert memory[24608] == 154

    final_data = {
        **pointers,
        "years": memory[694],
        "m": memory[8096],
        "amt": memory[16256],
        "firstnum": memory[18938],
        "slope": memory[20492],
        "k": memory[24608],
    }
    for name, value in final_data.items():
        print(f"{name:<9} = {value}")


if __name__ == "__main__":
    main()
